### ISSUE TITLE
Proposal: Forward SIGINT & SIGTERM

### ISSUE THREAD

**Issue by jclangst** [earthly/earthly#1167]
Earthly Version: `0.5.22`

## Current State
It seems that when the `earthly` process receives a SIGINT or SIGTERM, earthly does not send those signals to the running processes in Buildkit: https://github.com/earthly/earthly/blob/9e6b945ca043b47c9eccdc0f423e19ca4577a854/cmd/earthly/main.go#L183.

We are using earthly for running our infrastructure tools (e.g., terraform / terragrunt). Without receiving those signals, those processes leave external state locks in place that complicate working with the infrastructure.

## Proposal

 It would be ideal for earthly to forward those signals to the processes inside of Buildkit, wait for them to complete, and only forcibly shutdown if the timeout is exceeded. This would give these processes a chance to clean themselves up without requiring manual intervention.

**Comment by vladaionescu** [comment]
I think what happens under the hood is that the connection to buildkit gets dropped as a result of earthly exiting, and buildkit sees that as a canceled context. I think it `SIGKILL`s the underlying processes at that point (but not sure). Perhaps it's better if it `SIGTERM`s and waits 10-30s for them to gracefully exit before `SIGKILL`ing. If that process is then restarted by another earthly it may need to first wait for the previous process to exit before restarting.

**Comment by jclangst** [comment]
Yes, that behavior would be inline with a lot of stateful infrastructure tooling (e..g, terraform / terragrunt).

I'd recommend a 30 second default timeout. If its not a big lift, I'd also recommend adding a `--termination-grace-period-seconds` argument to the `RUN` step that could allow more granular configuration if desired. This would be fairly isomorphic to the Kubernetes behavior which I bet a lot of people using earthly are familiar with.

Regarding: 

 > If that process is then restarted by another earthly it may need to first wait for the previous process to exit before restarting.

I am not really sure about the Buildkit internals, but why would earthly need to wait for a previous process to exit before restarting? Anecdotally, I haven't had any issues running concurrent earthly processes on the same targets.

Unless you mean you want to provide a sort of built-in mutex? I wouldn't worry about that as most of the tools that require cleanup have their own state locking mechanisms.

**Comment by vladaionescu** [comment]
> I am not really sure about the Buildkit internals, but why would earthly need to wait for a previous process to exit before restarting?

By default, Buildkit would want to reuse computation. So if a process with the same hash is already running, it'll attach its outcome to it. So if it fails (because of a SIGTERM), we don't want that "cancelled" outcome to be propagated to the newer invocation. Somehow these two will need to be disconnected from each other.

**Comment by jclangst** [comment]
Interesting. I am assuming the behavior I experienced of not re-attaching was because I disabled caching for the infrastructure-related targets. I'd have to do a little bit more research to have a cogent suggestion as this is slightly beyond my current knowledge level of the buildkit internals.

**Comment by gpoulin** [comment]
Not sure if it's related but that seems to prevent clean up when running docker locally. For example
```
VERSION 0.6

server:
    FROM python:3.9-slim
    ENTRYPOINT python -m http.server
    SAVE IMAGE server

serve:
    LOCALLY
    WITH DOCKER --load=+server
        RUN docker run --rm -p 127.0.0.1:8000:8000 server
    END
```

If sending a SIGTERM after `earthly +serve` the container is still running.

**Comment by vladaionescu** [comment]
The use-case for `LOCALLY` makes sense - I think that's different enough for it to be a separate issue. Mind filing it separately @gpoulin?
